### PR TITLE
8342043: Split Opaque4Node into OpaqueTemplateAssertionPredicateNode and OpaqueNotNullNode

### PR DESCRIPTION
**Pull request by chhagedorn**
### Two Uses of `Opaque4`
The `Opaque4` node is currently used for two things:
1. Non-null checks of intrinsics to accompany a cast to non-null where we implicitly know that an object is non-null but C2 does not. This is required to fold control away if the cast becomes top at some point.
2. Template Assertion Predicates

### How to Differentiate between Uses
The only reliable way to differentiate between the two uses is to check if there are `OpaqueLoop*Nodes` above the `Opaque4` node which should only be found for a Template Assertion Predicate. This can be done with the method `assertion_predicate_has_loop_opaque_node()` which follows the inputs of the `Opaque4` recursively.

### Problems by Sharing `Opaque4` Nodes for Two Concepts
This sharing of the `Opaque4` comes with some problems:
- One need to be careful when checking for `Opaque4` nodes if the code can be applied for non-null checks and/or Template Assertion Predicates. Ideally, we should add assertions to exclude the unexpected case (mostly done).
- Walking the graph with `assertion_predicate_has_loop_opaque_node()` has some overhead, especially when checking the negation, we could walk quite some nodes in the worst case.
- There is no real benefit of (re)using `Opaque4` nodes for two concepts and could easily be separated.

### Split `Opaque4` into Two Classes to Separate Uses
Therefore, this patch proposes to split `Opaque4` into a new `OpaqueNonNull` and a `OpaqueTemplateAssertionPredicate` class (the latter accompanies the already split off `OpaqueInitializedAssertionPredicate` class). I went through all the uses and updated them accordingly.

As a consequence, I could turn `assertion_predicate_has_loop_opaque_node()` into a debug only method since it's now only used in assertion code. I additionally removed the second input of the `Opaque4` nodes since these nodes have always eventually been replaced by true.

### Turning `OpaqueTemplateAssertionPredicate` into a Non-Macro Node
The `Opaque4` was a macro node. The `OpaqueNotNull` still is but the `OpaqueTemplateAssertionPredicate` is turned into a non-macro node and is removed after loop opts. At  that point, loops can no longer be split and we do no need to create Initialized Assertion Predicates from the templates anymore. Thus, they can be cleaned up in the post loop opts IGVN.

### Pre-Requisite for Replacing Uncommon Traps with Halt Nodes for Template Assertion Predicates
Eventually, I want to get rid of UCTs for Template Assertion Predicates completely. They have been used out of convenience to reuse code in Loop Predication. But there is a problem when copying them to loops where we do not have traps (for example, for the remaining loop when peeling one iteration off or when copying them to the main loop). We cannot use UCTs anymore and need to fall back to `Halt` nodes.

Supporting the UCT and `Halt` node format for Template Assertion Predicates is difficult and does not really give us a benefit. Therefore, I want to get rid of UCTs and only use `Halt` nodes with [JDK-8342047](https://bugs.openjdk.org/browse/JDK-8342047). This patch lays the foundation to enable this change to still easily detect a Template Assertion Predicate.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342043](https://bugs.openjdk.org/browse/JDK-8342043): Split Opaque4Node into OpaqueTemplateAssertionPredicateNode and OpaqueNotNullNode (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21608/head:pull/21608` \
`$ git checkout pull/21608`

Update a local copy of the PR: \
`$ git checkout pull/21608` \
`$ git pull https://git.openjdk.org/jdk.git pull/21608/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21608`

View PR using the GUI difftool: \
`$ git pr show -t 21608`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21608.diff">https://git.openjdk.org/jdk/pull/21608.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21608#issuecomment-2426731304)

### CODE REVIEW

File: src/hotspot/share/opto/escape.cpp [hunk 1]
```diff
@@ -574,14 +574,14 @@ bool ConnectionGraph::can_reduce_check_users(Node* n, uint nesting) const {
...
-          // Bail out in such case - we need to preserve Opaque4 for correct
-          // processing predicates after loop opts.
+          // We may have an OpaqueNotNull node between If and Bool nodes. Bail out in such case.
           bool can_reduce = (iff->Opcode() == Op_If) && iff->in(1)->is_Bool() && iff->in(1)->in(1)->is_Cmp();
           if (can_reduce) {
             Node* iff_cmp = iff->in(1)->in(1);
             int opc = iff_cmp->Opcode();
             can_reduce = (opc == Op_CmpP || opc == Op_CmpN) && can_reduce_cmp(n, iff_cmp);
+          } else {
+            assert(iff->in(1)->is_OpaqueNotNull(), "must be OpaqueNotNull");
```
Comment (chhagedorn) [inline src/hotspot/share/opto/escape.cpp:584]: I don't think we can have a Template Assertion Predicate here, so I added an assertion. If this ever fails, we can just remove the assertion add a test for it.

File: src/hotspot/share/opto/loopopts.cpp [hunk 2]
```diff
@@ -2201,7 +2201,9 @@ void PhaseIdealLoop::clone_loop_handle_data_uses(Node* old, Node_List &old_new,
       // For example, it is unexpected that there is a Phi between an
       // AllocateArray node and its ValidLengthTest input that could cause
       // split if to break.
-      if (use->is_If() || use->is_CMove() || use->is_Opaque4() || use->is_OpaqueInitializedAssertionPredicate() ||
+      assert(!use->is_OpaqueTemplateAssertionPredicate(),
+             "should not clone a Template Assertion Predicate which should be removed once it's useless");
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopopts.cpp:2205]: `OpaqueTemplateAssertionPredicate` nodes should only have a single unique use. I therefore added this assert here.

File: src/hotspot/share/opto/macro.cpp [hunk 3]
```diff
@@ -2428,7 +2428,7 @@ void PhaseMacroExpand::eliminate_macro_nodes() {
         break;
       default:
         assert(n->Opcode() == Op_LoopLimit ||
-               n->is_Opaque4()             ||
+               n->is_OpaqueNotNull()       ||
```
Comment (chhagedorn) [inline src/hotspot/share/opto/macro.cpp:2431]: `OpaqueTemplateAssertionPredicate` is not a macro node.

File: src/hotspot/share/opto/predicates.cpp [hunk 4]
```diff
@@ -137,22 +137,19 @@ bool RuntimePredicate::is_predicate(Node* node, Deoptimization::DeoptReason deop
...
 bool TemplateAssertionPredicate::is_predicate(Node* node) {
   if (!may_be_assertion_predicate_if(node)) {
     return false;
   }
   IfNode* if_node = node->in(0)->as_If();
-  if (if_node->in(1)->is_Opaque4()) {
-    return RegularPredicateWithUCT::has_valid_uncommon_trap(node) || AssertionPredicateWithHalt::has_halt(node);
-  }
-  return false;
+  return if_node->in(1)->is_OpaqueTemplateAssertionPredicate();
```
Comment (chhagedorn) [inline src/hotspot/share/opto/predicates.cpp:146]: Detection is now easier. I plan to add more verification code that we also find the correct nodes on the failing path. But I'll wait with that until we have only Halt nodes for Template Assertion Predicates.

File: src/hotspot/share/opto/split_if.cpp [hunk 5]
```diff
@@ -322,7 +322,8 @@ bool PhaseIdealLoop::clone_cmp_down(Node* n, const Node* blk1, const Node* blk2)
           assert( bol->is_Bool(), "" );
           if (bol->outcnt() == 1) {
             Node* use = bol->unique_out();
-            if (use->is_Opaque4() || use->is_OpaqueInitializedAssertionPredicate()) {
+            if (use->is_OpaqueNotNull() || use->is_OpaqueTemplateAssertionPredicate() ||
+                use->is_OpaqueInitializedAssertionPredicate()) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/split_if.cpp:326]: Code is executed with both new opaque nodes (hit during testing). Same below.

**Comment by rwestrel** [comment]
> OpaqueTemplateAssertionPredicate is turned into a non-macro node and is removed after loop opts. At that point, loops can no longer be split and we do no need to create Initialized Assertion Predicates from the templates anymore. Thus, they can be cleaned up in the post loop opts IGVN.

The role of an assertion predicate is to catch an out of range input to a range check Cast or Conv nodes. If the `OpaqueTemplateAssertionPredicate` is removed after loop opts then you don't expect out of range values to appear after loop opts are over. But is it really the case? Why wouldn't an igvn later on not cause an out of range value? Also what's the benefit of changing when the opaque node is removed?

**Comment by chhagedorn** [comment]
> > OpaqueTemplateAssertionPredicate is turned into a non-macro node and is removed after loop opts. At that point, loops can no longer be split and we do no need to create Initialized Assertion Predicates from the templates anymore. Thus, they can be cleaned up in the post loop opts IGVN.
> 
> The role of an assertion predicate is to catch an out of range input to a range check Cast or Conv nodes. If the `OpaqueTemplateAssertionPredicate` is removed after loop opts then you don't expect out of range values to appear after loop opts are over. But is it really the case? Why wouldn't an igvn later on not cause an out of range value? Also what's the benefit of changing when the opaque node is removed?

The `OpaqueTemplateAssertionPredicate` is only used for a Template Assertion Predicate from which we create new Initialized Assertion Predicates from when splitting a loop during loop opts, for example for Loop Peeling. When we no longer split loops, we do not need to create new Initialized Assertion Predicates anymore. So, we replace the `OpaqueTemplateAssertionPredicate` nodes with true to let the Template Assertion Predicates be folded away in the post loop opts IGVN round. However, the `OpaqueInitializedAssertionPredicate` nodes, which accompany the Initialized Assertion Predicates and make sure control is also folded away, will be kept until macro node expansion.

**Comment by rwestrel** [comment]
> The `OpaqueTemplateAssertionPredicate` is only used for a Template Assertion Predicate from which we create new Initialized Assertion Predicates from when splitting a loop during loop opts, for example for Loop Peeling. When we no longer split loops, we do not need to create new Initialized Assertion Predicates anymore. So, we replace the `OpaqueTemplateAssertionPredicate` nodes with true to let the Template Assertion Predicates be folded away in the post loop opts IGVN round. However, the `OpaqueInitializedAssertionPredicate` nodes, which accompany the Initialized Assertion Predicates and make sure control is also folded away, will be kept and are only removed during macro node expansion (like the `Opaque4` nodes before).

That makes sense. Thanks for the explanation.

**Review by vnkozlov — APPROVED** [review]
Seems reasonable.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

File: src/hotspot/share/opto/loopTransform.cpp [hunk 6]
```diff
@@ -1922,25 +1925,21 @@ void PhaseIdealLoop::update_main_loop_assertion_predicates(Node* ctrl, CountedLo
...
     }
     Node* bol = iff->in(1);
-    if (bol->is_Opaque4()) {
-      if (assertion_predicate_has_loop_opaque_node(iff)) {
-        // This is a Template Assertion Predicate for the initial or last access.
-        // Create an Initialized Assertion Predicates for it accordingly:
-        // - For the initial access a[init] (same as before)
-        // - For the last access a[init+new_stride-orig_stride] (with the new unroll stride)
-        prev_proj = create_initialized_assertion_predicate(iff, init, max_value, prev_proj);
-      } else {
```
Comment (rwestrel) [inline src/hotspot/share/opto/loopTransform.cpp:1932]: Why is it safe to remove this?

Comment (chhagedorn) [reply src/hotspot/share/opto/loopTransform.cpp:1932]: Before the patch, we could have an `Opaque4` for an Template Assertion Predicate or Non-null check. Therefore, we additionally called `assertion_predicate_has_loop_opaque_node(iff)` to distinguish them. Now, it is enough to check for `bol->is_OpaqueTemplateAssertionPredicate()`. We can still have an `OpaqueNotNull` node which I now assert on L1942 instead.

**Review by rwestrel — APPROVED** [review]
Looks good to me.

**Comment by chhagedorn** [comment]
Thanks Roland for your review!

**Pull request closed without merging**